### PR TITLE
Implemented VirtualService and RouteTable based canary rollouts

### PR DESCRIPTION
**Pull request by dmitri-d**
This PR adds support for VirtualService- and RouteTable-based canary rollouts. Examples of possible configurations are in examples/ directory.

VirtualService-based canaries: a single VS, multiple routes configurations are supported.
RouteTable-based canaries: single or multiple RTs, mulitple route configurations are supported.
There's also an example of a blue-green deployment in examples/ dir; it's there for documentation purposes, no changes to the plugin were required.

Pertinent rollouts docs: https://argo-rollouts.readthedocs.io/en/stable/features/canary/#canary-deployment-strategy
Note that the plugin interface is minimal, only responsible for setting the weight, everything else is handled by the controller itself and is outside the scope of this PR.

To test/reproduce RT-based rollouts (assuming a k8s with Gloo Edge has been installed):
install Argo Rollouts:
`kubectl apply -k deploy/ --context=kind-kind`

initial setup that creates services, upstreams, etc:
`kubectl apply -f examples/canaries-with-multiple-routetables/setup.yaml --context=kind-kind`

create Argo rollout (creates a rollout, ReplicaSets, starts routing traffic to the stable rs):
`kubectl apply -f examples/canaries-with-multiple-routetables/create-rollout.yaml --context=kind-kind`

start forwarding traffic to the canary:
`kubectl apply -f examples/canaries-with-multiple-routetables/start-canary-traffic-routing.yaml --context=kind-kind`

To check the status of the rollout (requires argo plugin installed, see here: https://github.com/argoproj/argo-rollouts/blob/master/docs/installation.md#kubectl-plugin-installation):
`kubectl argo rollouts get rollout echo-rollouts -n echo -w --context=kind-kind`

To promote the rollout between steps:
`kubectl argo rollouts promote echo-rollouts -n echo --context kind-kind`

### CODE REVIEW

**Comment by dmitri-d** [comment]
* Added support for 'single' RouteAction destination

File: go.mod [hunk 1]
```diff
@@ -1,6 +1,6 @@
 module github.com/argoproj-labs/rollouts-plugin-trafficrouter-glooedge
 
-go 1.20
+go 1.19
```
Comment (kdorosh) [inline go.mod]: why go down?

Comment (dmitri-d) [reply go.mod]: no good reason. Fixed

File: pkg/plugin/plugin.go [hunk 2]
```diff
@@ -75,10 +82,23 @@ func (r *RpcPlugin) SetWeight(rollout *v1alpha1.Rollout, desiredWeight int32, ad
...
-	} else if rollout.Spec.Strategy.BlueGreen != nil {
-		return r.handleBlueGreen(rollout, glooPluginConfig)
+	if (glooPluginConfig.VirtualServiceSelector == nil && glooPluginConfig.RouteTableSelector == nil) ||
+		(glooPluginConfig.VirtualServiceSelector != nil && glooPluginConfig.RouteTableSelector != nil) {
+		return pluginTypes.RpcError{
+			ErrorString: "one of virtualService or routeTable selectors must be set in solo-io/glooedge plugin configuration",
+		}
+	}
+
+	if rollout.Spec.Strategy.Canary != nil && glooPluginConfig.VirtualServiceSelector != nil {
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go]: What does this nil guard for the `rollout.Spec.Strategy.Canary` provide us? Given that the earlier call to `getPluginConfig` unmarshals the `rollout.Spec.Strategy.Canary.TrafficRouting.Plugins[PluginName]` aren't we assured that the value is non-nil? If we aren't assured of this, should we add this nil check within the unmarshal step?

File: pkg/plugin/plugin.go [hunk 3]
```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
 	return &glooplatformConfig, nil
 }
 
-func (r *RpcPlugin) getRouteTables(ctx context.Context, rollout *v1alpha1.Rollout, pluginConfig *GlooEdgeTrafficRouting) ([]*GlooMatchedRouteTable, error) {
-	if pluginConfig.RouteTableSelector == nil {
-		return nil, fmt.Errorf("routeTable selector is required")
-	}
+func getStableServiceName(rollout *v1alpha1.Rollout) string {
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go:152]: I like these!

File: pkg/plugin/virtual_service_canary.go [hunk 4]
```diff
@@ -0,0 +1,96 @@
...
+	gwv1 "github.com/solo-io/solo-apis/pkg/api/gateway.solo.io/v1"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+func (r *RpcPlugin) handleCanaryUsingVirtualService(
+	ctx context.Context,
+	rollout *v1alpha1.Rollout,
+	desiredWeight int32,
+	additionalDestinations []v1alpha1.WeightDestination,
```
Comment (sam-heilbron) [inline pkg/plugin/virtual_service_canary.go]: additionalDestinations is unused in this method, can we remove it, but more importantly can you help me understand why its not needed?

File: pkg/plugin/virtual_service_canary.go [hunk 5]
```diff
@@ -0,0 +1,96 @@
...
+	pluginConfig *GlooEdgeTrafficRouting) error {
+
+	vs, err := r.getVirtualService(ctx, rollout, pluginConfig)
+	if err != nil {
+		return err
+	}
+
+	originalVs := &gwv1.VirtualService{}
+	vs.DeepCopyInto(originalVs)
```
Comment (sam-heilbron) [inline pkg/plugin/virtual_service_canary.go:25]: Why is this deepDopy necessary? Currently we read the VS based on the name/namespace, so we should have a fresh object and then we deep copy it as well, which feels like an unnecessary duplicate step, but perhaps I'm misunderstanding something

File: pkg/plugin/virtual_service_canary.go [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+
+	return nil
+}
+
+func (r *RpcPlugin) getDestinationsInVirtualService(
+	rollout *v1alpha1.Rollout,
+	pluginConfig *GlooEdgeTrafficRouting,
+	vs *gwv1.VirtualService) (ret []destinationPair, error error) {
+
+	if vs.Spec.GetVirtualHost() == nil || vs.Spec.GetVirtualHost().GetRoutes() == nil {
```
Comment (sam-heilbron) [inline pkg/plugin/virtual_service_canary.go]: nit: We can consolidate this check to just:
```
vs.Spec.GetVirtualHost().GetRoutes() == nil
```

The getter methods that we create for our APIs have nil guards built into them: https://github.com/solo-io/solo-apis/blob/gloo-main/pkg/api/gateway.solo.io/v1/virtual_service.pb.go#L410

File: pkg/plugin/plugin.go (hunk 2 above)
Comment (sam-heilbron) [reply pkg/plugin/plugin.go]: As I look at https://argo-rollouts.readthedocs.io/en/stable/features/canary/#trafficrouting, it seems like the TrafficRouting object is not guaranteed to be non-nil, am I understanding that correctly?

File: pkg/plugin/plugin.go [hunk 7]
```diff
@@ -28,7 +29,9 @@ type RpcPlugin struct {
 }
 
 type GlooEdgeTrafficRouting struct {
-	RouteTableSelector *DumbObjectSelector `json:"routeTableSelector" protobuf:"bytes,1,name=routeTableSelector"`
+	RouteTableSelector     *DumbObjectSelector `json:"routeTable" protobuf:"bytes,1,name=routeTable"`
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go]: nit: I think some comments on these fields would help clarify the responsibility of each object. For example, since `Routes` is used so frequently, it's not immediately clear what those represent. 

Also, I noticed that our VirutalServiceSelector re-uses the DumpObjectSelector (which makes sense), but doesn't actually use the labels configuration that is exposed. I imagine there are technical reasons behind this decision but it's hard to follow the breadcrumbs without being closely familiar with the product. Could you either clarify this on the type definition here, or in the code that consumes this to clarify that it's intentional the labels aren't used and not an oversight?

File: pkg/plugin/plugin.go [hunk 8]
```diff
@@ -66,7 +69,11 @@ func (r *RpcPlugin) UpdateHash(rollout *v1alpha1.Rollout, canaryHash, stableHash
 	return pluginTypes.RpcError{}
 }
 
-func (r *RpcPlugin) SetWeight(rollout *v1alpha1.Rollout, desiredWeight int32, additionalDestinations []v1alpha1.WeightDestination) pluginTypes.RpcError {
+func (r *RpcPlugin) SetWeight(
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go:77]: Referring to the code just above, why are we safe to not implement UpdateHash? From the comment on the function:
```
// UpdateHash informs a traffic routing reconciler about new canary, stable, and additionalDestination(s) pod hashes
```

it's not immediately clear how that is used by Argo

File: pkg/plugin/virtual_service_canary.go [hunk 9]
```diff
@@ -0,0 +1,96 @@
...
+
+	originalVs := &gwv1.VirtualService{}
+	vs.DeepCopyInto(originalVs)
+
+	allDestinations, err := r.getDestinationsInVirtualService(rollout, pluginConfig, vs)
+	if err != nil {
+		return err
+	}
+
+	r.maybeConvertSingleToMulti([]routeTableWithDestinations{{Destinations: allDestinations}})
```
Comment (sam-heilbron) [inline pkg/plugin/virtual_service_canary.go:32]: I'm a little confused by the usage of routeTableWithDestinations in this method to handle VirtualServices. We don't pass a RouteTable into that function, and seem to work to convert a list of single destinations into multi destinations. Since destinations have a reference to the `DestinationParent` RouteAction, why do we need a reference to the RouteTable as well?

File: pkg/plugin/plugin.go [hunk 10]
```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
...
-	}
+func getCanaryServiceName(rollout *v1alpha1.Rollout) string {
+	return rollout.Spec.Strategy.Canary.CanaryService
+}
 
-	var rts []*gloov1.RouteTable
+func (r *RpcPlugin) maybeConvertSingleToMulti(routeTables []routeTableWithDestinations) {
+	for i := range routeTables {
+		for j := range routeTables[i].Destinations {
+			if routeTables[i].Destinations[j].DestinationsParent.GetMulti() != nil {
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go:163]: In this loop, we could iterate over (destinationIndex, destination), and instead of looking up the object by the index again, just use the reference in the range. Is there a reason we don't do that?

File: pkg/plugin/plugin.go [hunk 11]
```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
...
-		r.LogCtx.Debugf("getRouteTables using ns:name ref %s:%s to get single table", pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace)
-		result, err := r.Client.RouteTables().GetRouteTable(ctx, pluginConfig.RouteTableSelector.Name, pluginConfig.RouteTableSelector.Namespace)
-		if err != nil {
-			return nil, err
+func (r *RpcPlugin) maybeCreateCanaryDestinations(
+	routeTables []routeTableWithDestinations, canaryName string) {
+
+	for i := range routeTables {
+		for j := range routeTables[i].Destinations {
+			if routeTables[i].Destinations[j].Canary != nil {
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go:182]: What would cause this case, where the Canary on the destination has already been set? Perhaps a comment indicating why we think its a valid case, and why we don't want to "overwrite" the previous definition

File: pkg/plugin/plugin.go [hunk 12]
```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
...
+		}
 
-		rts, err = r.Client.RouteTables().ListRouteTables(ctx, opts)
-		if err != nil {
-			return nil, err
+		if route.GetRouteAction().GetMulti().GetDestinations() == nil {
+			continue
 		}
-		r.LogCtx.Debugf("getRouteTables listing tables with opts %+v; found %d routeTables", opts, len(rts))
+		ret = append(ret, r.getDestinationsInMulti(route, rollout)...)
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go]: nit: I think some of this logic can be simplified. It seems like this logic attempts to do the following:

For each route:
- if a list of available routes is provided, the route name must be in the list
- if the route singleAction is defined, append the WeightedDestinations from that route
- If the route MultiActions is defined, append the WeightedDestinations
- continue

I think what you have is good as is, it just took some extra for me to undersatnd the logic and I figured, the following might be easier to reason through:

```go
  if route.GetRouteAction().GetSingle() != nil {
    ret = append(ret, r.getDestinationInSingle(route, rollout)...)
    continue
  }

  if route.GetRouteAction().GetMulti() != nil {
    ret = append(ret, r.getDestinationsInMulti(route, rollout)...)
    continue
  }
```

File: pkg/plugin/plugin.go [hunk 13]
```diff
@@ -116,142 +136,119 @@ func getPluginConfig(rollout *v1alpha1.Rollout) (*GlooEdgeTrafficRouting, error)
...
+	return ret
+}
 
-	for _, rt := range rts {
-		matchedRt := &GlooMatchedRouteTable{
-			RouteTable: rt,
+func (r *RpcPlugin) getDestinationsInMulti(route *gwv1.Route, rollout *v1alpha1.Rollout) (ret []destinationPair) {
+	var stable, canary *v1.WeightedDestination
+	for _, dst := range route.GetRouteAction().GetMulti().GetDestinations() {
+		if dst.GetDestination() == nil || dst.GetDestination().GetUpstream() == nil ||
```
Comment (sam-heilbron) [inline pkg/plugin/plugin.go]: nit: Same as some other comments, the double nil check I think this unnecessary, since `dst.GetDestination().GetUpstream()` will not panic on a nil destination

File: pkg/plugin/route_table_canary.go [hunk 14]
```diff
@@ -0,0 +1,123 @@
...
+
+	if pluginConfig.RouteTableSelector.Name == "" && len(pluginConfig.RouteTableSelector.Labels) == 0 {
+		return nil, fmt.Errorf("name or labels field must be set in RouteTable selector")
+	}
+
+	if pluginConfig.RouteTableSelector.Name != "" {
+		return r.getRouteTable(ctx, namespace, pluginConfig.RouteTableSelector.Name)
+	}
+
+	return r.listRouteTables(ctx, namespace, pluginConfig)
```
Comment (sam-heilbron) [inline pkg/plugin/route_table_canary.go:96]: Why do we support label selections for RouteTables but not for VirtualServices?

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (sam-heilbron) [reply pkg/plugin/plugin.go:152]: I noticed on the API definition for these names, the fields are optional. Do we need to validate that the field is set anywhere?

**Review by sam-heilbron — COMMENTED** [review]
Looking great! A few comments/questions

File: pkg/plugin/route_table_canary.go (hunk 14 above)
Comment (dmitri-d) [reply pkg/plugin/route_table_canary.go:96]: the gist is we support a single VS but multiple RTs; This is an MVP, figured that it's sufficient to support just multiple RTs. Multiple VS support can be added if/when customers ask for it.

File: pkg/plugin/virtual_service_canary.go (hunk 4 above)
Comment (dmitri-d) [reply pkg/plugin/virtual_service_canary.go]: It's passed from Argo to SetWeight() and contains a list of destinations that argo is going to update. The only piece of data we don't already have is PodTemplateHash and we don't use it in our plugin, rendering the whole thing not useful to us.

File: pkg/plugin/virtual_service_canary.go (hunk 5 above)
Comment (dmitri-d) [reply pkg/plugin/virtual_service_canary.go:25]: This is needed to generate a patch later: https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-glooedge/pull/1/files#diff-ab569cd7f37c9fd5ca9dfdeda525bc8fd68af3442955c6165e067321e1a942fbR42

File: pkg/plugin/virtual_service_canary.go (hunk 9 above)
Comment (dmitri-d) [reply pkg/plugin/virtual_service_canary.go:32]: Probably a bad name for this struct; needed a common-enough object to contain destinations and parent containers for them. RouteTable is specific to RT-based canaries and not used here.

File: pkg/plugin/virtual_service_canary.go (hunk 6 above)
Comment (dmitri-d) [reply pkg/plugin/virtual_service_canary.go]: fixed.

File: pkg/plugin/plugin.go (hunk 7 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go]: The plugin currently supports rollouts using a single VS but multiple RTs; it was decided that we can add support for multiple VS' if/when customers request it.

File: pkg/plugin/plugin.go (hunk 8 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go:77]: Not entire sure either. I *think* it's only used to update hashes in `additionalDestinations` that we don't use. None of the other plugins I looked at implement it, and its lack didn't appear to affect rollouts.

File: pkg/plugin/plugin.go (hunk 2 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go]: Removed it, I don't think the plugin will ever be called for non-canary rollouts.

> it seems like the TrafficRouting object is not guaranteed to be non-nil, am I understanding that correctly?

Hmm, yeah, it's a possibility, added a nil check to `getPluginConfig()`

File: pkg/plugin/plugin.go (hunk 3 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go:152]: Added a validation and a test

File: pkg/plugin/plugin.go (hunk 10 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go:163]: We need tho modify the original a bit later (here:[maybeConvertSingleToMulti](https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-glooedge/pull/1/files#diff-fc137d62147eb6a9c89c0566479ff219320ca691f2505631ebacbe513bcc8108R153)), I thought it might be confusing to mix both styles. I can change it if you think it's ok.

File: pkg/plugin/plugin.go (hunk 11 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go:182]: It's possible the the VS/RTs routes already have a canary destination defined, there's no point re-creating it, it's going to be the same. Added a comment.

File: pkg/plugin/plugin.go (hunk 12 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go]: Haha, yeah, I see what you mean, changed.

File: pkg/plugin/plugin.go (hunk 13 above)
Comment (dmitri-d) [reply pkg/plugin/plugin.go]: fixed.

File: README.md [hunk 15]
```diff
@@ -1,5 +1,92 @@
...
+        multi:
+          destinations:
+          - destination:
+              upstream:
+                name: echo-v1
+                namespace: echo
+            weight: 100
+```
+
+Note that only `multi` routeActions are supported. It's ok to define a destination for a stable release only. The names for stable and canary `Upstream`s are expected to match the name of the services (and `stableService` and `canaryService` fields of the plugin configuration).
```
Comment (bdlilley) [inline README.md]: You added support for single destinations also right?

File: pkg/plugin/plugin.go (hunk 8 above)
Comment (bdlilley) [reply pkg/plugin/plugin.go:77]: I think it's only used if the plugin adds entirely new routes "owned" by the plugin; we have no need to do that so prob can ignore this like the other plugins do

File: pkg/plugin/route_table_canary.go (hunk 14 above)
Comment (bdlilley) [reply pkg/plugin/route_table_canary.go:96]: agree w/ @dmitri-d , I can see possible use cases but we can wait till customer asks

**Review by bdlilley — CHANGES_REQUESTED** [review]
Only minor thing is top level readme says only multi destinations are supported

File: README.md (hunk 15 above)
Comment (dmitri-d) [reply README.md]: Yes, see here: https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-glooedge/pull/1/files#diff-56d28ecf3c4e8c775d9ec794e963fde93acc1b78444376ebb8a8a9a7a9b0c415R30 and here: https://github.com/argoproj-labs/rollouts-plugin-trafficrouter-glooedge/pull/1/files#diff-ab569cd7f37c9fd5ca9dfdeda525bc8fd68af3442955c6165e067321e1a942fbR33

**Review by sam-heilbron — APPROVED** [review]
For some reason I'm unable to "resolve" conversations that I opened, and have now been addressed. But LGTM 🚀

**Review by bdlilley — APPROVED** [review]
